### ISSUE TITLE
Cowboy Hard Caches Static File Output Unless I Nuke Everything

### ISSUE THREAD

**Issue by ajm113** [ninenines/cowboy#1248]
This is a bit of a very unusual issue for myself. I don't know if I have something setup wrong, or if there is black magic in the works, or I drank too many fizzy drinks to read the docs slow enough, but I'll do my best to describe my issue...

So I have gulp converting my pug.js input into a friendly browser readable .html file. Whenever I make any changes to the pug.js file for gulp to update the .html file in my priv folder it updates perfectly fine. No issue here. However, when I attempt to see my changes on the server using Cowboy, I don't see my changes reflect.

Here's what I've done, running `rebar3 shell`

- Attempt to Google my issue. I haven't seen anyone run into this, and I am using the most recent version. So perhaps this is a new bug, and I'm the only one thus far who has ran into it?
- Delete the priv folder of my app's output _build. -- No dice
- Blow away browser cache. -- Nothing
- Directly input my priv folder path using this `{"/assets/[...]", cowboy_static, {dir, code:priv_dir(bombinaid) ++ "/assets"}}`. -- Nope.avi
- Directly edit the HTML file inside the priv folder. -- Negative!
- Also I notice if I remove all the text inside the html file parts of the old file still appear. So for some reason Cowboy registers the removed/added text, but still shows old files or garbage data.
- Blow away _build folder all together. IT WORKS! But why do I have to nuke everything to see my changes?

TL;DR ver: Cowboy doesn't recognize local file changes to update the server, unless I blow away cowboy.

My rebar.config
```
{erl_opts, [debug_info]}.
{shell, [{apps, [
    bombinaid,
    kernel,
    stdlib,
    cowboy
]}]}.

{deps, [
	{cowboy, {git,"https://github.com/ninenines/cowboy.git", {tag, "2.2.0"}}}
]}.

{relx, [{release, { bombinaid, "0.0.1" },
         [bombinaid,
          sasl,
          cowboy
          ]},

        {dev_mode, true},
        {include_erts, true},

        {extended_start_script, true}]
}.

```

My repo having the issue: https://github.com/ajm113/bombinaid.com

**Comment by ferd** [comment]
If you check the paths in `_build` you'll see that they should be symlinks:

```
λ bombinaid.com → master → ls -al _build/default/lib/bombinaid/
total 16
drwxrwxr-x 4 ferd ferd 4096 Dec 23 21:17 .
drwxrwxr-x 6 ferd ferd 4096 Dec 23 21:17 ..
drwxrwxr-x 2 ferd ferd 4096 Dec 23 21:17 ebin
lrwxrwxrwx 1 ferd ferd   19 Dec 23 21:17 include -> ../../../../include
lrwxrwxrwx 1 ferd ferd   16 Dec 23 21:17 priv -> ../../../../priv
drwxrwxr-x 2 ferd ferd 4096 Dec 23 21:17 .rebar3
lrwxrwxrwx 1 ferd ferd   15 Dec 23 21:17 src -> ../../../../src
```

If you blow that directory away, the files still exist in the top-level `priv/` directory. Deleting it may just mean that it gets re-linked from the top-level one, and similarly for the manual input with priv_dir. That being said, if that was relevant, chances are that blowing `_build` away wouldn't have any other impact either.

You can try a more hardcore setup for it by going `rebar3 as prod release` (which disables dev_mode and cancels out symlinks) try the same experiment from there. This should let you see if symlinking could in any way be to blame for what you're seeing.

That being said, I tried with your repo to see the problem, and there is nothing in the priv/ directory to test with.

**Comment by ajm113** [comment]
@ferd Yes, that priv directory is intentionally empty since Gulp/Node, generate the output HTML file. You can throw in a `index.html` for some quick testing. 

I'll give the prod a try, and simply try to streamline my changes back/front-end only if it works. I'll ping back when I try some more things out.

I should mention that the project is also running in a Vagrant box, w/ erlang/rebar that's file sharing to the host machine. idk if that would affect things too, but I thought I should throw that out if that's also what might be causing the headaches. Again, I'll ping back sometime tonight or tomorrow of what I find out.

**Comment by ajm113** [comment]
@ferd Seems like when I installed the project directly on my host machine it runs fine. For whatever reason symlinks simply don't play nice on VMs that are also running on shared to the host. -- Even when you are running commands/programs directly on the VM. Wonder if it's because Linux Mint vs Ubuntu treats linking slightly differently perhaps?

**Comment by ajm113** [comment]
Closing this issue, since it doesn't relate to Cowboy. More of a OS/environment issue.

**Comment by essen** [comment]
There are a lot of issues with files on Virtualbox shared folders. Don't use them. Instead setup the VM standalone and SSH or similar (sshfs should work perfectly).

**Comment by ajm113** [comment]
I'll have to give sshfs a try. I was just thinking of Samba, but even Samba is a pain when it comes to text editors or IDE sometimes. -- Would think most text editing applications would support network drives perfectly fine...
